docs: add explanation about installing a slice with rockcraft (#148)
canonical/rockcraft: docs/how-to/install-slice.rst

You can now build your ROCK:

.. code-block:: sh

    $ rockcraft

The output will be:

.. code-block:: text

    Launching instance...
    Retrieved base bare for amd64
    Extracted bare:latest
    Executed: pull build-context
    Executed: pull pebble
    Executed: overlay build-context
    Executed: overlay pebble
    Executed: build build-context
    Executed: build pebble
    Executed: stage build-context
    Executed: stage pebble
    Executed: prime build-context
    Executed: prime pebble
    Executed parts lifecycle
    Created new layer
    Labels and annotations set to ['org.opencontainers.image.version=0.0.1', 'org.opencontainers.image.title=ourrocks', 'org.opencontainers.image.ref.name=ourrocks', 'org.opencontainers.image.licenses=GPL-3.0', 'org.opencontainers.image.created=2022-11-25T14:00:47.470814+00:00', 'org.opencontainers.image.base.digest=c4d1cae85485fb5bf8483a440f7e47b0fd2592ff114117cd4763604fbf6ae7a4']
    Exported to OCI archive 'ourrocks_0.0.1_amd64.rock'

You can test that the OpenSSL binaries have been correctly installed with the following:

.. code-block:: sh

>     $ skopeo --insecure-policy copy oci-archive:ourrocks_0.0.1_amd64.rock docker-daemon:chisel-openssl:latest

Author: Done

    $ docker run chisel-openssl openssl

The skopeo output will be:

.. code-block:: text

    Getting image source signatures
    Copying blob 253d707d7e97 done
    Copying blob 7044a53e1935 done
    Copying config c114b59704 done
    Writing manifest to image destination
    Storing signatures

The output of the docker command will be OpenSSL's default help message:

.. code-block:: sh

    help:

    Standard commands
    asn1parse         ca                ciphers           cmp
    cms               crl               crl2pkcs7         dgst
    dhparam           dsa               dsaparam          ec
    ecparam           enc               engine            errstr
    fipsinstall       gendsa            genpkey           genrsa
    help              info              kdf               list
    mac               nseq              ocsp              passwd
    pkcs12            pkcs7             pkcs8             pkey
    pkeyparam         pkeyutl           prime             rand
    rehash            req               rsa               rsautl
    s_client          s_server          s_time            sess_id

>     <... many more lines of output>

Reviewer: ```suggestion
    <... many more lines of output>

And that's it! You've now built your own ROCK from a custom Chisel release. Next step: share your slice definitions file with others!
```

> How to install your own package slice ?

Reviewer 2: ```suggestion
How to install your own package slice
```

To keep it consistent with the "slice creation" how-to

Let's assume you want to install the OpenSSL binaries slice created in :ref:`here <create-slice>`...

**First**, clone the Chisel releases repository:

.. code-block:: sh

    # Let's assume we are working with Ubuntu 22.04
    $ git clone -b ubuntu-22.04 https://github.com/canonical/chisel-releases/

This repository acts as the database of slice definitions files for each Chisel release (Chisel releases are named analogously to Ubuntu releases, and mapped into Git branches within the repository).

Chisel will only recognize slices belonging to a Chisel release, so you need to copy your slice definitions file - ``openssl.yaml`` in this example - into the ``chisel-releases/slices`` folder. Note that if a slice definitions file with the same name already exists, it most likely means that the package you're slicing has already been sliced before, and in this case, you only need to merge your changes into that existing file.

At this point, you should be able to find your custom OpenSSL slice ``bins`` in the local Chisel release:

.. code-block:: sh

    grep -q "bins" chisel-releases/slices/openssl.yaml && echo "My slice exists"

> If you wanted to test it with Chisel alone, you could now simply run ``chisel cut --release ./chisel-releases --root . openssl_bins``. You should end up with a few folders, amongst which there would be ``./usr/bin/openssl``.

Reviewer 2: might be a good idea to use a different `--root`, to avoid polluting the current working directory

**To install the custom package slice into a ROCK though**, you need to use Rockcraft!

Start by initializing a new Rockcraft project:

.. code-block:: sh

    $ rockcraft init

> after this command, you should find a new ``rockcraft.yaml`` file in your current path.

Reviewer 2: ```suggestion
After this command, you should find a new ``rockcraft.yaml`` file in your current path.
```

    Executed: stage pebble
    Executed: prime build-context
    Executed: prime pebble
    Executed parts lifecycle
    Created new layer
    Labels and annotations set to ['org.opencontainers.image.version=0.0.1', 'org.opencontainers.image.title=custom-openssl-rock', 'org.opencontainers.image.ref.name=custom-openssl-rock', 'org.opencontainers.image.licenses=GPL-3.0', 'org.opencontainers.image.created=2022-11-25T14:00:47.470814+00:00', 'org.opencontainers.image.base.digest=c4d1cae85485fb5bf8483a440f7e47b0fd2592ff114117cd4763604fbf6ae7a4']
    Exported to OCI archive 'custom-openssl-rock_0.0.1_amd64.rock'

You can test that the OpenSSL binaries have been correctly installed with the following:

.. code-block:: sh

    $ skopeo --insecure-policy copy oci-archive:custom-openssl-rock_0.0.1_amd64.rock docker-daemon:chisel-openssl:latest

The output will be:

.. code-block:: text

    Getting image source signatures
    Copying blob 253d707d7e97 done
    Copying blob 7044a53e1935 done
    Copying config c114b59704 done
    Writing manifest to image destination
    Storing signatures

And after:

.. code-block:: sh

    $ docker run chisel-openssl openssl

> The output of the docker command will be OpenSSL's default help message:

Reviewer: ```suggestion
The output of the Docker command will be OpenSSL's default help message:
```

> .. code-block:: sh

Reviewer: ```suggestion
.. code-block:: text
```

At this point, you should be able to find your custom OpenSSL slice ``bins`` in the local Chisel release:

.. code-block:: sh

    grep -q "bins" slices/openssl.yaml && echo "My slice exists"

If you wanted to test it with Chisel alone, you could now simply run ``chisel cut --release ./chisel-releases --root . openssl_bins``. You should end up with a few folders, amongst which there would be ``./usr/bin/openssl``.

**To install the custom package slice into a ROCK though**, you need to use Rockcraft!

Start by initializing a new Rockcraft project:

.. code-block:: sh

    $ rockcraft init

after this command, you should find a new ``rockcraft.yaml`` file in your current path.

Now you need to copy and paste this code into the ``rockcraft.yaml`` file (feel free to adjust the metadata, but pay special attention to the ``parts`` section):

> .. code-block:: yaml

Reviewer: previous comment unresolved

Author: it's had been add i don't see the issue here

Author: can i resolve this one or not ?

Reviewer: u can but pls unresolve https://github.com/canonical/rockcraft/pull/148#discussion_r1031646261

Author: Done

> How-to install from my own custom slice definitions?

Reviewer: ```suggestion
How to install your own package slices?
```

****************************************************

> First, you need to clone this repository:

Reviewer: ```suggestion
When a specific package slice isn't available on the `upstream Chisel releases <https://github.com/canonical/chisel-releases>`_, you'll more likely end up creating your own slice definition.

Once you have it though, the most obvious question is: _how can I install this custom slice with Chisel?_

Let's then assume you want to install the OpenSSL binaries slice created in :ref:`here <create-slice>`...

**First**, you need to clone the Chisel releases repository:
```

.. code-block:: sh

>     $ git clone https://github.com/canonical/chisel-releases/

Reviewer: ```suggestion
    # Let's assume we are working with Ubuntu 22.04
    $ git clone -b ubuntu-22.04 https://github.com/canonical/chisel-releases/
```

> This repository is the database of slices definitions file. After that, you need to put your openssl.yaml file into the chisel-releases/slices folder.

Reviewer: ```suggestion
This repository acts as the database of slice definitions files for each Chisel release (Chisel releases are named analogously to Ubuntu releases, and mapped into Git branches within the repository). 

Chisel will only recognize slices belonging to a Chisel release, so you need to copy your slice definitions file - ``openssl.yaml`` in this example - into the ``chisel-releases/slices`` folder. Note that if a slice definitions file with the same name already exists, it most likely means that the package you're slicing has already been sliced before, and in this case, you only need to merge your changes into that existing file.
```

> | Now we will install our openssl slice into our ROCK'S, so to initialize a rock you need to do:

Reviewer: ```suggestion
At this point, you should be able to find your custom OpenSSL slice ``bins`` in the local Chisel release:

.. code-block:: sh
  
    grep -q "bins" slices/openssl.yaml && echo "My slice exists"    

If you wanted to test it with Chisel alone, you could now simply run ``chisel cut --release ./chisel-releases --root . openssl_bins``. You should end up with a few folders, amongst which there would be ``./usr/bin/openssl``.

**To install the custom package slice into a ROCK though**, you need to use Rockcraft! 

Start by initializing a new Rockcraft project:
```

.. code-block:: sh

    $ rockcraft init

> after this command your folder looks like this:

Reviewer: ```suggestion
after this command, you should find a new ``rockcraft.yaml`` file in your current path.
```

> .. code-block:: text

Reviewer: u can delete this code block

    -chisel-releases/
    -rockcraft.yaml

> Now you need to copy and paste this code into the rockcraft file:

Reviewer: ```suggestion
Now you need to copy and paste this code into the ``rockcraft.yaml`` file (feel free to adjust the metadata, but pay special attention to the ``parts`` section):
```

> .. code-block:: yaml

Reviewer: pls rename it. it is a ROCK, not a slice. (same for summary and description)

Also, with the `organize` I suspect the chisel-releases will end up in the build, and that's undesired.

Author: even without, i add a rm -rf to delete the things we don't want

Reviewer: you could do that, but that's not what we should advocate. That's not a good practice, nor is it exploiting our tool's potential. I.e., we can split the things that are needed to the build phase, vs the ones that are needed to be primed into the final image

Author: What do you recommend not using dump or ?

Reviewer: you need that plugin, but you don't want to prime anything from it

Author: I don't see how to say's sent the chisel folder but remove it after without cleanup or override

Reviewer: there must be a way. just needs some digging through the snapcraft docs

Author: We can use this : https://snapcraft-utils-library.readthedocs.io/en/latest/lib/cleanup.html

Author: I have push a working solution : 

For the dump plugins: 
* I sent the file into the stage area thanks to keyword stage
* I remove the file from the prime area thanks to the keyword prime
For chisel:
* I get the database and slices folder thanks to the CRAFT_STAGE variable

I test it and at the end the slices folder and chisel.yaml is not inside the rocks.

output of 

```sh
$ ls $CRAFT_STAGE && ls $CRAFT_PRIME
```
this command was put inside the override prime of install-slice

```
2022-11-25 01:13:44.444 :: 2022-11-25 00:13:43.800 :: + ls /root/stage         
2022-11-25 01:13:44.444 :: 2022-11-25 00:13:43.801 :: chisel.yaml              
2022-11-25 01:13:44.444 :: 2022-11-25 00:13:43.801 :: slices                   
2022-11-25 01:13:44.444 :: 2022-11-25 00:13:43.801 :: usr                      
2022-11-25 01:13:44.444 :: 2022-11-25 00:13:43.802 :: + ls /root/prime         
2022-11-25 01:13:44.445 :: 2022-11-25 00:13:43.802 :: etc                      
2022-11-25 01:13:44.445 :: 2022-11-25 00:13:43.802 :: lib                      
2022-11-25 01:13:44.445 :: 2022-11-25 00:13:43.803 :: lib64                    
2022-11-25 01:13:44.445 :: 2022-11-25 00:13:43.803 :: usr  
```

Author: We can even do something like this:

```yaml
    name: ourrocks
    base: bare
    build_base: "ubuntu:22.04"
    version: '0.0.1'
    summary: tutorial rocks
    description: |
        Rocks with inside our slice
    license: GPL-3.0
    platforms:
        amd64:
    parts:
        build-context:
            plugin: dump
            source: chisel-releases/
            source-type: local
            stage:
                - chisel.yaml
                - slices
            override-prime: chisel cut --release $CRAFT_STAGE/ --root . openssl_bins
```

```
2022-11-25 01:27:04.288 :: 2022-11-25 00:26:47.075 :: + ls /root/stage                                               
2022-11-25 01:27:04.288 :: 2022-11-25 00:26:47.076 :: chisel.yaml                                                    
2022-11-25 01:27:04.288 :: 2022-11-25 00:26:47.076 :: slices                                                         
2022-11-25 01:27:04.288 :: 2022-11-25 00:26:47.076 :: usr   
2022-11-25 01:27:04.289 :: 2022-11-25 00:27:03.583 :: + ls /root/prime                                               
2022-11-25 01:27:04.289 :: 2022-11-25 00:27:03.584 :: etc                                                            
2022-11-25 01:27:04.289 :: 2022-11-25 00:27:03.585 :: lib                                                            
2022-11-25 01:27:04.289 :: 2022-11-25 00:27:03.585 :: lib64                                                          
2022-11-25 01:27:04.289 :: 2022-11-25 00:27:03.585 :: usr    
```

I'll let you choose the best way, and I voted for the first way because it's easier to understand, in my opinion.

Reviewer: this is looking much better now. thanks.

I'd just like to get @tigarmo 's feedback on this last YAML. @tigarmo do you see an even cleaner and more intuitive way of doing this?

Reviewer 2: Hi friends. Well done on the debugging journey ;). Some notes:

- The only different between the `nil` and the `dump` plugins is that `dump` stages all source files by default. So if you only want to use the source files (`chisel-release/`, in this case) during build it's often simpler to just use `nil`.
- You usually want to avoid doing any kind of "big" building work during `prime`, and in this case it's possible to use the build stage to cut the slices. I tried the following part declaration here and it seems to work:

```yaml
parts:
    build-context: 
        plugin: nil 
        source: chisel-releases/
        source-type: local
        override-build: 
            chisel cut --release ./ --root ${CRAFT_PART_INSTALL} openssl_bins
```

With the `nil` plugin the actual contents of the `chisel-releases/` dir isn't included in the final artifact by default, so it's just a matter of having chisel cut the slices into the correct directory. I would say that this is the more "idiomatic" way of doing this process.
Hope this helps!

Reviewer 2: Not sure if this has been linked already, but [these](https://craft-parts.readthedocs.io/en/latest/parts_steps.html?highlight=CRAFT_PART_SRC%20#step-execution-environment) are the environment variables that you can access during the build

Author: test and this solution work, right now I have pushed this solution

    name: ourslice
    base: bare
    build_base: "ubuntu:22.04"
    version: '0.0.1'
    summary: Install the slice
    description: |
        Slice install tutorial
    license: GPL-3.0
    platforms:
        amd64:
    parts:
        build-context:
            plugin: dump
            source: chisel-releases/
            source-type: local
            organize:
                '*': chisel-releases/

>         installslice:

Reviewer: ```suggestion
        install-slice:
```

            plugin: go

>             source: https://github.com/canonical/chisel

Reviewer: do you actually need to compile Chisel? it is already being shipped with the Rockcraft Snap, so I wonder whether it could be already available at build time...

Author: I don't think it's available at the build name, do you have the command for testing?

Reviewer: @tigarmo since utilities like Skopeo and Chisel are shipped with the Rockcraft snap, is there a way to use them from within a part?

Author: Done, only need to call chisel. I have updated the yaml file

            source-type: git
            build-snaps:
                - go/1.17/stable
            after: [build-context]
            override-prime: $CRAFT_PART_INSTALL/bin/chisel cut --release chisel-releases/ --root . openssl_bins # name of your slice

> The build-context permits us to send our chisel-releases folder to the builder. The install slice section enables us to install our slice thanks to chisel into the ROCK.

Reviewer: ```suggestion
The "build-context" part allows you to send the local ``chisel-releases`` folder into the builder. The "install-slice" section enables you to install your custom slice.
```

> if you need more information about the yaml file, you can check the :ref:`reference` section.

Reviewer: ```suggestion
Please not that this level of customization is only needed when you want to install from a custom Chisel release. If the desired slice definitions are already upstream, then you can simply use ``stage-packages``, as demonstrated in :ref:`here <chisel-example>`. 
```

> | To build your ROCK, you need to do the following:

Reviewer: ```suggestion
You can now build your ROCK:
```

.. code-block:: sh

    $ rockcraft

> 

Reviewer: please paste the output

> To run it on docker, you need to convert the rocks into a docker image:

Reviewer: ```suggestion
```

> .. code-block:: sh

Reviewer: ```suggestion
```

    $ skopeo --insecure-policy copy oci-archive:ourslice_0.0.1_amd64.rock docker-daemon:chisel-openssl:latest

> You can test it with the following:

Reviewer: ```suggestion
You can test that the OpenSSL binaries have been correctly installed with the following:
```

.. code-block:: sh

>     $ docker run chisel-openssl openssl

Reviewer: ```suggestion
    $ skopeo --insecure-policy copy oci-archive:ourslice_0.0.1_amd64.rock docker-daemon:chisel-openssl:latest
    $ docker run chisel-openssl openssl
```

Reviewer: also, please paste the output of the commands and fix the oci archive filename once you change the YAML above

> The output will be OpenSSL's default help message, which starts like this:

Reviewer: ```suggestion
The output will be OpenSSL's default help message:
```
